File: storage/storage.go
package storage

import (
	"context"
	"io"
	"time"
)

// ObjectInfo holds metadata about a stored object.
type ObjectInfo struct {
	Key  string
	Size int64
}

// ObjectStorage defines the interface for cloud storage operations.
type ObjectStorage interface {
	// Upload sends data to the given object key.
	Upload(ctx context.Context, key string, reader io.Reader, size int64) error

	// Head returns the size of the object in bytes.
	// If the object does not exist, it returns -1 and nil error.
	Head(ctx context.Context, key string) (size int64, err error)

	// ListObjects returns all objects whose key starts with the given prefix.
	ListObjects(ctx context.Context, prefix string) ([]ObjectInfo, error)

	// DeleteObjects deletes all objects with the given keys.
	DeleteObjects(ctx context.Context, keys []string) error

	// PresignedURL generates a time-limited download URL for the given object key.
	PresignedURL(ctx context.Context, key string, expires time.Duration) (string, error)
}

File: apps/ship/src/storage/storage.ts
export interface ObjectInfo {
  key: string;
  size: number;
}

export interface ObjectStorage {
  /** Upload the local file at filePath to key. */
  upload(key: string, filePath: string): Promise<void>;
  /** Object size in bytes, or -1 when the object does not exist. */
  head(key: string): Promise<number>;
  listObjects(prefix: string): Promise<ObjectInfo[]>;
  deleteObjects(keys: string[]): Promise<void>;
  presignedUrl(key: string, expiresSeconds: number): Promise<string>;
}

export const DELETE_BATCH = 1000;

export function chunk<T>(items: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < items.length; i += size) {
    out.push(items.slice(i, i + size));
  }
  return out;
}
